feat: OffensiveInput and HumorInput API implementation

artikandri opened the pull request:
Features added with this PR: 
- all changes related to OffensiveInput and HumorInput will be saved on input change 

What's changed: 
- `frontend/components/tasks/affectiveAnnotation/humor/HumorInput.vue`: connecting the API, fixed ts errors
- `frontend/components/tasks/affectiveAnnotation/offensive/OffensiveInput.vue`: connecting the API, fixed ts errors 
- `frontend/i18n/*/projects/annotation.js`: added missing translations
- `frontend/pages/projects/_id/affective-annotation/index.vue`: fixed conflicts, integrated HumorInput and OffensiveInput

Notable: 
- some orders may have changed because of the conflicts, sorry for that XD   
- I added `ts-ignore` to ignore the typescript errors caused by accessing an object with string, since I havent found a better solution XD   Probably will fix it later.

teddy-f-47 commented on frontend/i18n/zh/projects/annotation.js:
```diff
@@ -13,16 +13,16 @@ export default {
   key: '键',
   value: '值',
   newText: '下一条',
-  articleTitle: '文章标题',
+  articleTitle: '文章标Question',
   articleId: '文章 ID',
   labelTypes: '标签类型',
```
I think the field `articleTitle` here is accidentally changed?

artikandri replied:
Lol I missed this, thank you for pointing it out XD  Alright

artikandri replied:
2. Could you please add the JSON files of the scale labels that will be imported for Offensive and Humor modes into frontend/static/formats/affective_annotation? I think it will be useful to have the files containing the expected scale labels in the same place, as we assume the scale labels to be 'fixed'.

Sure, I will XD

teddy-f-47 reviewed (commented):
Hello, thank you for your great work! I think in general the code looks fine. There are just a few comments:
(1) There is a minor comment below about one of the translation files being changed (?).
(2) Could you please add the JSON files of the scale labels that will be imported for Offensive and Humor modes into `frontend/static/formats/affective_annotation`? I think it will be useful to have the files containing the expected scale labels in the same place, as we assume the scale labels to be 'fixed'.
(3) - removed -
(4) It is pretty obvious that our codes look quite different XD But I think your approach in general is very good. So, after this PR is merged, I will try to refactor the Summary-Emotions-Others modes to be more-or-less similar to your approach. But I still need more time to do that, and temporarily imo it is okay to leave the code as it is, as the most important thing is to have the app working 😅 

Edit : I think the issue I mentioned in number 3 is because I inputted bad scale labels. So I removed it. I think I will wait for the scale labels files to be added before testing it again. XD

haanh764 reviewed (commented):
Thank you for your great implementation! So far the code looks fine, but I have some small comment (not really comment but question XD)

(1) I have checked the save data in table `labels_textlabel` and the `question` field is something like this `subquestion3.substatement8##Poglądy polityczne, partie`. Personally I think it's better to save the question only without the `subquestion3.substatement8` part. So how do you think?

(2) In the section `(4) Nature of humor. It makes you laugh because:`
Since it's the checkboxes, so now I see that you're saving the text as `subquestion4.substatement11##1`, and the question as `subquestion4.substatement11##Kontrast`. So still the same comment as (1) about the `subquestion4.substatement11##` shoud not be saved, and I think it's better here if you save the text field as the selected option, and the question will be the question of the whole section `(4) Nature of humor. It makes you laugh because`, So how do you think? XD

artikandri commented:
> Thank you for your great implementation! So far the code looks fine, but I have some small comment (not really comment but question XD)
> 
> (1) I have checked the save data in table `labels_textlabel` and the `question` field is something like this `subquestion3.substatement8##Poglądy polityczne, partie`. Personally I think it's better to save the question only without the `subquestion3.substatement8` part. So how do you think?
> 
> (2) In the section `(4) Nature of humor. It makes you laugh because:` Since it's the checkboxes, so now I see that you're saving the text as `subquestion4.substatement11##1`, and the question as `subquestion4.substatement11##Kontrast`. So still the same comment as (1) about the `subquestion4.substatement11##` shoud not be saved, and I think it's better here if you save the text field as the selected option, and the question will be the question of the whole section `(4) Nature of humor. It makes you laugh because`, So how do you think? XD

1. Ah, I added the keys because it will make the mapping a lot easier XD   I can save it as the question only but it will take more time (and possibly can cause more errors) for me to map the question answers to the form data XD  So can I save the keys?
2. Same answer XD  Now I removed `##1` from the answer (it was a workaround because previously the system won't accept duplicated values) and save the question as the value. I personally think it will be better to save '1' or 'Checked' as the answer tho, but any string will do. 
![image](https://user-images.githubusercontent.com/12537724/197486843-0f538d00-2390-453d-a919-7afab7f2c208.png)

artikandri commented:
> Thank you for your great implementation! So far the code looks fine, but I have some small comment (not really comment but question XD)
> 
> (1) I have checked the save data in table `labels_textlabel` and the `question` field is something like this `subquestion3.substatement8##Poglądy polityczne, partie`. Personally I think it's better to save the question only without the `subquestion3.substatement8` part. So how do you think?
> 
> (2) In the section `(4) Nature of humor. It makes you laugh because:` Since it's the checkboxes, so now I see that you're saving the text as `subquestion4.substatement11##1`, and the question as `subquestion4.substatement11##Kontrast`. So still the same comment as (1) about the `subquestion4.substatement11##` shoud not be saved, and I think it's better here if you save the text field as the selected option, and the question will be the question of the whole section `(4) Nature of humor. It makes you laugh because`, So how do you think? XD

Nevermind; I removed the key from the question now XD  You can check it. Then again, it's more prone to errors since now we're using string as the key and it might be slower as well since I have to add 2 additional loop to get the key XD 

![image](https://user-images.githubusercontent.com/12537724/197490262-ac6c2b4c-77ae-4246-a335-b6a1d9fcfe35.png)

teddy-f-47 reviewed (approved):
Hello, I have tested it and everything works correctly on my machine. I can confirm that the issue on point 3 on my initial comment (which was removed) was caused by importing incorrect scale labels, because now I have tested using the scale labels files provided in the frontend directory, and everything works perfectly.

For the mapping mechanism, I think this is acceptable for now. I understand your worries about the code being error-prone, but since the question value is taken from the `this.polishAnnotation` from i18n directory, I think it should be fine. Of course this relies on the assumption that the `polishAnnotation` is fixed and won't change.

Alternatively, you can hard-code the `question` value in the frontend code (in both HumorInput.vue and OffensiveInput.vue). This will be "safer" in case something changes in the i18n translation directory. For example, if the question in i18n is changed from "What is your name?" to "What is your name???", this will cause (minor) problems because previous annotations won't be loaded properly. And then, when the user enters a new value, we will now have duplicate entries in the db, the first one with "What is your name?" question and the second one with the "What is your name???" question. Meanwhile, if the `question` value is hard-coded, any change in i18n directory will not affect the functionality. This approach's perspective is that the translation string is in a totally separate world from the `question` value that is saved into db. I am using this approach for the Summary mode and Others mode. But, then again, since we are assuming the question strings as fixed and won't change, I think your current code is totally valid and fine, so you don't need to change it if you disagree with my suggestion. XD

Meanwhile, with the scale data, the mapping relies on the imported scale label file; so, as long as the correct file (with the correct text writing) is imported, it should be fine. This is the reason why I think working on the auto-import for the scale labels needs to be done as soon as possible.

Finally, thank you for the excellent work! I approve this for merging, as I said the suggestion above is optional. XD

artikandri commented:
> Hello, I have tested it and everything works correctly on my machine. I can confirm that the issue on point 3 on my initial comment (which was removed) was caused by importing incorrect scale labels, because now I have tested using the scale labels files provided in the frontend directory, and everything works perfectly.
> 
> For the mapping mechanism, I think this is acceptable for now. I understand your worries about the code being error-prone, but since the question value is taken from the `this.polishAnnotation` from i18n directory, I think it should be fine. Of course this relies on the assumption that the `polishAnnotation` is fixed and won't change.
> 
> Alternatively, you can hard-code the `question` value in the frontend code (in both HumorInput.vue and OffensiveInput.vue). This will be "safer" in case something changes in the i18n translation directory. For example, if the question in i18n is changed from "What is your name?" to "What is your name???", this will cause (minor) problems because previous annotations won't be loaded properly. And then, when the user enters a new value, we will now have duplicate entries in the db, the first one with "What is your name?" question and the second one with the "What is your name???" question. Meanwhile, if the `question` value is hard-coded, any change in i18n directory will not affect the functionality. This approach's perspective is that the translation string is in a totally separate world from the `question` value that is saved into db. I am using this approach for the Summary mode and Others mode. But, then again, since we are assuming the question strings as fixed and won't change, I think your current code is totally valid and fine, so you don't need to change it if you disagree with my suggestion. XD
> 
> Meanwhile, with the scale data, the mapping relies on the imported scale label file; so, as long as the correct file (with the correct text writing) is imported, it should be fine. This is the reason why I think working on the auto-import for the scale labels needs to be done as soon as possible.
> 
> Finally, thank you for the excellent work! I approve this for merging, as I said the suggestion above is optional. XD

1. Thank you for the suggestion XD However I personally will avoid hard-coded code as much as possible XD  So I will stick to using the annotation file. The current mechanism requires the question to be related to the answer anyway, and the question is related to the annotation file. If it comes to that, I will import an additional json file perhaps to store the labels. 
2. Thank you for the confirmation of using the wrong labels XD I will fix it in the next PR.

haanh764 reviewed (approved):
Thank you so much! I think everything looks oke now! Approved XD
Only one very minor suggestion as I said in point (2) that for the question `4.Nature of humor. It makes you laugh because:`, now the saved data is "text": "Niecenzuralne słowa", "question": "Niecenzuralne słowa" (they're the same and I understand why), therefore I think it maybe good if the question could be "4.Nature of humor. It makes you laugh because:" (?). It will be much easier in retrieve the data later (ofc if it's possible, if not then it's oke). And as I said this is just suggestion you can consider or not XD

artikandri commented:
> Thank you so much! I think everything looks oke now! Approved XD Only one very minor suggestion as I said in point (2) that for the question `4.Nature of humor. It makes you laugh because:`, now the saved data is "text": "Niecenzuralne słowa", "question": "Niecenzuralne słowa" (they're the same and I understand why), therefore I think it maybe good if the question could be "4.Nature of humor. It makes you laugh because:" (?). It will be much easier in retrieve the data later (ofc if it's possible, if not then it's oke). And as I said this is just suggestion you can consider or not XD

Oh i got it now, sorry xD  my mistake, I missed your point. Its possible but that way it will be harder (more complicated) for me to map. Anyway, I will take that suggestion and work on it later, thank you!
